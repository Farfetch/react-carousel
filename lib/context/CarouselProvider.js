"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ = require("./");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var CarouselProvider = /*#__PURE__*/function (_Component) {
  _inherits(CarouselProvider, _Component);

  var _super = _createSuper(CarouselProvider);

  _createClass(CarouselProvider, null, [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var isRTL = props.isRTL,
          itemsToShow = props.itemsToShow,
          isInfinite = props.isInfinite,
          itemsToScroll = props.itemsToScroll;
      return _objectSpread(_objectSpread({}, state), {}, {
        isRTL: isRTL,
        itemsToShow: itemsToShow,
        isInfinite: isInfinite,
        itemsToScroll: itemsToScroll
      });
    }
  }]);

  function CarouselProvider(props) {
    var _this;

    _classCallCheck(this, CarouselProvider);

    _this = _super.call(this, props);
    _this._setItemsLength = _this._setItemsLength.bind(_assertThisInitialized(_this));
    _this._setIsMovementBlocked = _this._setIsMovementBlocked.bind(_assertThisInitialized(_this));
    _this._goTo = _this._goTo.bind(_assertThisInitialized(_this));
    _this._goNext = _this._goNext.bind(_assertThisInitialized(_this));
    _this._goPrev = _this._goPrev.bind(_assertThisInitialized(_this));
    _this.state = {
      activeItem: _this.props.startItem,
      isRTL: false,
      isInfinite: false,
      isMovementBlocked: false,
      itemsToShow: 1,
      itemsLength: 0,
      setItemsLength: _this._setItemsLength,
      setIsMovementBlocked: _this._setIsMovementBlocked,
      goTo: _this._goTo,
      goNext: _this._goNext,
      goPrev: _this._goPrev,
      itemsToScroll: 1,
      direction: null
    };
    return _this;
  }

  _createClass(CarouselProvider, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.startItem !== this.props.startItem) {
        this.setState({
          activeItem: this.props.startItem
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return /*#__PURE__*/_react["default"].createElement(_.CarouselContext.Provider, {
        value: this.state
      }, children);
    }
  }, {
    key: "_getDirection",
    value: function _getDirection(newActiveItem) {
      var activeItem = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.state.activeItem;
      var lastItem = this.state.itemsLength - 1;

      if (newActiveItem === 0 && lastItem === activeItem) {
        return 'next';
      }

      if (activeItem === 0 && newActiveItem === lastItem) {
        return 'prev';
      }

      return activeItem < newActiveItem ? 'next' : 'prev';
    }
  }, {
    key: "_setActiveItem",
    value: function _setActiveItem(newActiveItem) {
      var _this2 = this;

      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var onAfterChange = this.props.onAfterChange;
      var activeItem = this.state.activeItem;

      if (newActiveItem === activeItem) {
        return null;
      }

      this.setState({
        activeItem: newActiveItem
      }, function () {
        return onAfterChange && onAfterChange(_objectSpread({
          index: newActiveItem,
          dir: _this2._getDirection(newActiveItem, activeItem)
        }, options));
      });
    }
  }, {
    key: "_setItemsLength",
    value: function _setItemsLength(itemsLength) {
      this.setState({
        itemsLength: itemsLength
      });
    }
  }, {
    key: "_setIsMovementBlocked",
    value: function _setIsMovementBlocked(isBlocked) {
      this.setState({
        isMovementBlocked: isBlocked
      });
    }
  }, {
    key: "_goTo",
    value: function _goTo(index) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var _this$state = this.state,
          itemsLength = _this$state.itemsLength,
          itemsToShow = _this$state.itemsToShow,
          isMovementBlocked = _this$state.isMovementBlocked;
      var isInvalidIndex = typeof index !== 'number' || itemsToShow > itemsLength || index < 0 || index > itemsLength - 1;

      if (isMovementBlocked || isInvalidIndex) {
        return null;
      }

      this._setActiveItem(index, options);
    }
  }, {
    key: "_goNext",
    value: function _goNext() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _this$state2 = this.state,
          activeItem = _this$state2.activeItem,
          isInfinite = _this$state2.isInfinite,
          itemsLength = _this$state2.itemsLength;
      var itemsToScroll = this.props.itemsToScroll;
      var newActiveItem = activeItem + itemsToScroll; // Stop at the first positon

      if (isInfinite && newActiveItem > itemsLength - 1) {
        newActiveItem = 0;
      }

      this.setState({
        direction: 'next'
      });

      this._goTo(newActiveItem, options);
    }
  }, {
    key: "_goPrev",
    value: function _goPrev() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _this$state3 = this.state,
          activeItem = _this$state3.activeItem,
          isInfinite = _this$state3.isInfinite,
          itemsLength = _this$state3.itemsLength;
      var itemsToScroll = this.props.itemsToScroll;
      var newActiveItem = activeItem - itemsToScroll; // Stop at the first positon

      if (isInfinite && newActiveItem < 0 && activeItem === 0) {
        newActiveItem = itemsLength - itemsToScroll;
      } // Move to a slide before the first position


      if (isInfinite && newActiveItem < 0 && activeItem !== 0) {
        newActiveItem = 0;
      }

      this.setState({
        direction: 'prev'
      });

      this._goTo(newActiveItem, options);
    }
  }]);

  return CarouselProvider;
}(_react.Component);

CarouselProvider.defaultProps = {
  itemsToShow: 1,
  isRTL: false,
  isInfinite: false,
  itemsToScroll: 1,
  startItem: 0
};
CarouselProvider.propTypes = {
  children: _propTypes["default"].node.isRequired,
  isRTL: _propTypes["default"].bool,
  isInfinite: _propTypes["default"].bool,
  itemsToShow: _propTypes["default"].number,
  onAfterChange: _propTypes["default"].func,
  itemsToScroll: _propTypes["default"].number,
  startItem: _propTypes["default"].number
};
var _default = CarouselProvider;
exports["default"] = _default;
module.exports = exports.default;