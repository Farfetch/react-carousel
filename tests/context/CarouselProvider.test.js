import { createSerializer } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import CarouselProvider from '../../src/context/CarouselProvider';
import React from 'react';

const MockComponent = jest.fn(() => null);

describe('<CarouselProvider/>', () => {
    beforeAll(() => {
        expect.addSnapshotSerializer(createSerializer());
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should render with required props', () => {
        const wrapper = shallow(
            <CarouselProvider>
                <MockComponent />
            </CarouselProvider>
        );
        const instance = wrapper.instance();

        const expectedContext = {
            activeItem: 0,
            isRTL: false,
            isInfinite: false,
            isMovementBlocked: false,
            itemsToShow: 1,
            itemsLength: 0,
            setItemsLength: instance._setItemsLength,
            setIsMovementBlocked: instance._setIsMovementBlocked,
            goTo: instance._goTo,
            goNext: instance._goNext,
            goPrev: instance._goPrev,
            direction: null,
            itemsToScroll: 1,
        };

        const { containerRef, ...currentValue } = wrapper.prop('value');

        expect(wrapper).toMatchSnapshot();
        expect(containerRef).not.toBeNull();
        expect(currentValue).toEqual(expectedContext);
    });

    test('should set the context correctly based on the props', () => {
        const wrapper = shallow(
            <CarouselProvider isInfinite itemsToShow={2} startItem={4} isRTL>
                <MockComponent />
            </CarouselProvider>
        );

        const instance = wrapper.instance();

        const expectedContext = {
            activeItem: 4,
            isRTL: true,
            isInfinite: true,
            isMovementBlocked: false,
            itemsToShow: 2,
            itemsLength: 0,
            setItemsLength: instance._setItemsLength,
            setIsMovementBlocked: instance._setIsMovementBlocked,
            goTo: instance._goTo,
            goNext: instance._goNext,
            goPrev: instance._goPrev,
            direction: null,
            itemsToScroll: 1,
        };

        const { containerRef, ...currentValue } = wrapper.prop('value');

        expect(containerRef).not.toBeNull();
        expect(currentValue).toEqual(expectedContext);
    });

    describe('_setItemsLength()', () => {
        test('should change state with item length', () => {
            const spySetState = jest.spyOn(
                CarouselProvider.prototype,
                'setState'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            expect(wrapper.state().itemsLength).toBe(0);

            instance._setItemsLength(1);
            expect(spySetState).toHaveBeenCalledTimes(1);
            expect(wrapper.state().itemsLength).toBe(1);
        });
    });

    describe('_setIsMovementBlocked()', () => {
        test('should change state with the new value', () => {
            const spySetState = jest.spyOn(
                CarouselProvider.prototype,
                'setState'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            expect(wrapper.state().isMovementBlocked).toBe(false);

            instance._setIsMovementBlocked(true);
            expect(spySetState).toHaveBeenCalledTimes(1);
            expect(wrapper.state().isMovementBlocked).toBe(true);
        });
    });

    describe('_goNext()', () => {
        test('should advance an item active', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goNext();
            expect(spyGoTo).toHaveBeenCalledWith(1, {});
        });

        test('should advance number of itemsToScroll from the active item', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider
                    isInfinite
                    itemsToShow={4}
                    itemsToScroll={4}
                    isRTL
                >
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 0 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goNext();
            expect(spyGoTo).toHaveBeenCalledWith(4, {});
        });

        test('should advance by itemsToScroll when slidesToShow equals itemsToScroll', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider isInfinite itemsToShow={4} itemsToScroll={4}>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 0 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goNext();
            expect(spyGoTo).toHaveBeenCalledWith(4, {});
        });

        test('should advance to the first item if new active item is bigger then the number of items', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider
                    isInfinite
                    itemsToShow={4}
                    itemsToScroll={4}
                    isRTL
                >
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 4 });

            const instance = wrapper.instance();

            instance._setItemsLength(8);
            instance._goNext();
            expect(spyGoTo).toHaveBeenCalledWith(0, {});
        });

        test('should advance an item active with aditional options', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goNext({ dir: 'right' });
            expect(spyGoTo).toHaveBeenCalledWith(1, { dir: 'right' });
        });

        test('should reset active item if advance to first item when is infinite', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider isInfinite>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(2);
            instance._goNext();
            instance._goNext();
            expect(spyGoTo).toHaveBeenLastCalledWith(0, {});
        });
    });

    describe('_goPrev()', () => {
        test('should subtract one from the active item', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goPrev();
            expect(spyGoTo).toHaveBeenCalledWith(0, {});
        });

        test('should retreat number of itemsToScroll from the active item', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider
                    isInfinite
                    itemsToShow={4}
                    itemsToScroll={4}
                    isRTL
                >
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 4 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goPrev();
            expect(spyGoTo).toHaveBeenCalledWith(0, {});
        });

        test('should retreat by itemsToScroll when slidesToShow equals itemsToScroll', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider isInfinite itemsToShow={4} itemsToScroll={4}>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 4 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goPrev();
            expect(spyGoTo).toHaveBeenCalledWith(0, {});
        });

        test('should advance an item active with aditional options', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goPrev({ dir: 'left' });
            expect(spyGoTo).toHaveBeenCalledWith(0, { dir: 'left' });
        });

        test('should put active item in last position if advance to last item when is infinite', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider isInfinite>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(2);
            instance._goPrev();
            expect(spyGoTo).toHaveBeenLastCalledWith(0, {});
        });

        test('should put active item as the first item when the activeItem is 0 and newActiveItem less than 0', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider isInfinite itemsToShow={4} itemsToScroll={4}>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(6);
            instance._goPrev();
            expect(spyGoTo).toHaveBeenLastCalledWith(0, {});
        });

        test('should put active item as the itemsLenght - itemsToScroll when activeItem is 0 and the newActiveItem is less than 0', () => {
            const spyGoTo = jest.spyOn(CarouselProvider.prototype, '_goTo');

            const wrapper = shallow(
                <CarouselProvider isInfinite itemsToShow={4} itemsToScroll={4}>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 0 });

            const instance = wrapper.instance();

            instance._setItemsLength(6);
            instance._goPrev();
            expect(spyGoTo).toHaveBeenLastCalledWith(2, {});
        });
    });

    describe('_goTo()', () => {
        test('should change active item if index is valid', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goTo(3);
            expect(spySetActiveItem).toHaveBeenCalledWith(3, {});
        });

        test('should change active item if index is valid and send options', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._goTo(3, { dir: 'left' });
            expect(spySetActiveItem).toHaveBeenCalledWith(3, { dir: 'left' });
        });

        test('should return null if the movement is blocked', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ isMovementBlocked: true });

            const instance = wrapper.instance();

            const returnedValue = instance._goTo(3);

            expect(returnedValue).toBeNull();
            expect(spySetActiveItem).toHaveBeenCalledTimes(0);
        });

        test('should return null if the sent value is not a number', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            const returnedValue = instance._goTo('3');

            expect(returnedValue).toBeNull();
            expect(spySetActiveItem).toHaveBeenCalledTimes(0);
        });

        test('should return null if value is lower than 0', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            const returnedValue = instance._goTo(-1);

            expect(returnedValue).toBeNull();
            expect(spySetActiveItem).toHaveBeenCalledTimes(0);
        });

        test('should return null if value is greater than number of items', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            wrapper.setState({ activeItem: 1 });

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            const returnedValue = instance._goTo(6);

            expect(returnedValue).toBeNull();
            expect(spySetActiveItem).toHaveBeenCalledTimes(0);
        });

        test('should return null if items to show are greater to number of items', () => {
            const spySetActiveItem = jest.spyOn(
                CarouselProvider.prototype,
                '_setActiveItem'
            );

            const wrapper = shallow(
                <CarouselProvider itemsToShow={2}>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(1);
            const returnedValue = instance._goTo(6);

            expect(returnedValue).toBeNull();
            expect(spySetActiveItem).toHaveBeenCalledTimes(0);
        });
    });

    describe('_setActiveItem()', () => {
        test('should change active item', () => {
            const spySetState = jest.spyOn(
                CarouselProvider.prototype,
                'setState'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setActiveItem(1);

            expect(spySetState).toHaveBeenCalledTimes(1);
            expect(wrapper.state().activeItem).toBe(1);
        });

        test('should return null if change to the same active item', () => {
            const spySetState = jest.spyOn(
                CarouselProvider.prototype,
                'setState'
            );

            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            const result = instance._setActiveItem(0);

            expect(spySetState).toHaveBeenCalledTimes(0);
            expect(wrapper.state().activeItem).toBe(0);
            expect(result).toBeNull();
        });

        test('should call callback if exist', () => {
            const spySetState = jest.spyOn(
                CarouselProvider.prototype,
                'setState'
            );
            const mockAfterChange = jest.fn();

            const wrapper = shallow(
                <CarouselProvider onAfterChange={mockAfterChange}>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(5);
            instance._setActiveItem(1);

            const expectedAfterChange = {
                index: 1,
                dir: 'next',
            };

            expect(spySetState).toHaveBeenCalledTimes(2);
            expect(mockAfterChange).toBeCalledTimes(1);
            expect(mockAfterChange).toHaveBeenCalledWith(expectedAfterChange);
            expect(wrapper.state().activeItem).toBe(1);
        });

        test('should call callback if exist, and pass all options', () => {
            const spySetState = jest.spyOn(
                CarouselProvider.prototype,
                'setState'
            );
            const mockAfterChange = jest.fn();

            const wrapper = shallow(
                <CarouselProvider onAfterChange={mockAfterChange}>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            const options = {
                dir: 'left',
                data: 'test',
            };

            instance._setItemsLength(5);
            instance._setActiveItem(1, options);

            const expectedAfterChange = {
                index: 1,
                ...options,
            };

            const callAfterChange = mockAfterChange.mock.calls[0][0];

            expect(spySetState).toHaveBeenCalledTimes(2);
            expect(mockAfterChange).toBeCalledTimes(1);
            expect(callAfterChange).toEqual(expectedAfterChange);
            expect(wrapper.state().activeItem).toBe(1);
        });
    });

    describe('_getDirection()', () => {
        const nextDirection = 'next';
        const prevDirection = 'prev';

        test('should return next if new active item is higher than the previous', () => {
            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(3);

            const result = instance._getDirection(1);

            expect(result).toBe(nextDirection);
        });

        test('should return prev if new active item is lower than the previous', () => {
            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(3);
            instance._setActiveItem(2);

            const result = instance._getDirection(1);

            expect(result).toBe(prevDirection);
        });

        test('should return prev if new active item is the last item of items', () => {
            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(2);
            instance._setActiveItem(0);

            const result = instance._getDirection(1);

            expect(result).toBe(prevDirection);
        });

        test('should return next if new active item is the first item of items', () => {
            const wrapper = shallow(
                <CarouselProvider>
                    <MockComponent />
                </CarouselProvider>
            );

            const instance = wrapper.instance();

            instance._setItemsLength(2);
            instance._setActiveItem(1);

            const result = instance._getDirection(0);

            expect(result).toBe(nextDirection);
        });
    });

    test('should update active item when startItem prop changes', () => {
        const wrapper = shallow(
            <CarouselProvider>
                <MockComponent />
            </CarouselProvider>
        );

        const prevProps = wrapper.instance().props;

        wrapper.setProps({ startItem: 0 });
        wrapper.instance().componentDidUpdate(prevProps);

        expect(wrapper.state('activeItem')).toBe(0);

        wrapper.setProps({ startItem: 5 });
        wrapper.instance().componentDidUpdate(prevProps);

        expect(wrapper.state('activeItem')).toBe(5);
    });
});
