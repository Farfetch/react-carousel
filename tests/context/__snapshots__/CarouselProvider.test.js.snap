// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarouselProvider/> should render with required props 1`] = `
<ContextProvider
  value={
    Object {
      "activeItem": 0,
      "containerRef": Object {
        "current": null,
      },
      "direction": null,
      "goNext": [Function],
      "goPrev": [Function],
      "goTo": [Function],
      "isInfinite": false,
      "isMovementBlocked": false,
      "isRTL": false,
      "itemsLength": 0,
      "itemsToScroll": 1,
      "itemsToShow": 1,
      "setIsMovementBlocked": [Function],
      "setItemsLength": [Function],
    }
  }
>
  <mockConstructor />
</ContextProvider>
`;
